### PR TITLE
Fix build & migrate MultipleFilesWatcher to Apache commons I/O

### PR DESCRIPTION
**Pull request by ctrlcctrlv**

### CODE REVIEW

File: pgptool-gui/pom.xml
Comment (ctrlcctrlv) [inline pgptool-gui/pom.xml:184]: This change was made to fix the direct error in #284.

File: pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java
Comment (ctrlcctrlv) [inline pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:42]: All these changes were made as your dependency resolution is broken and you had the TODO anyway so even fixing above error is not enough.

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.3.1:compile (default-compile) on project gui: Compilation failure: Compilation failure: 
[ERROR] /home/fred/Workspace/pgptool/pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:[42,52] error: cannot find symbol
[ERROR] 
[ERROR]  package ru.skarpushin.swingpm.modelprops.virtualtable
[ERROR] /home/fred/Workspace/pgptool/pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:[59,9] error: cannot find symbol
[ERROR] 
[ERROR]  class MultipleFilesWatcher
[ERROR] /home/fred/Workspace/pgptool/pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:[149,5] error: cannot find symbol
[ERROR] 
[ERROR]  class BaseFolder
[ERROR] /home/fred/Workspace/pgptool/pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:[151,38] error: cannot find symbol
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

File: pgptool-gui/installer/launch4j.xml
Comment (ctrlcctrlv) [inline pgptool-gui/installer/launch4j.xml:6]: This PR is now ready for review as I figured out what you were doing wrong with the Windows cross compile. Obviously I have no `/home/sergeyk` on my system :-)

**Comment by skarpushin** [comment]
@ctrlcctrlv hi! Thank you very much for your contribution! This is actually first notable contribution to this project. 
I'll try to find time next week to review & test your suggested changes.

There are couple things that I don't know how to handle though, any thoughts are appreciated:

1. I.e. what If I want to adjust something in this PR, how do I do this? Do I checkout this PR, modify code and push it? Or there is some other way? (never handled PRs on Github before)

1. You want to add your name to license prefix for the file (Copyright (C) 2023 Fredrick R. Brennan <copypaste@kittens.ph>) but this prefix generated uniformly for all files by plugin that I use in IDE (Eclipse). So even if I accept this change next time your name will disappear. Would you be satisfied if I move mentioning of your name to class comment so it will be retained there?

**Comment by ctrlcctrlv** [comment]
> You want to add your name to license prefix for the file (Copyright (C) 2023 Fredrick R. Brennan [copypaste@kittens.ph](mailto:copypaste@kittens.ph)) but this prefix generated uniformly for all files by plugin that I use in IDE (Eclipse). So even if I accept this change next time your name will disappear. Would you be satisfied if I move mentioning of your name to class comment so it will be retained there?

I don't mind at all, I assumed you might have abandoned this project actually so only added myself in anticipation of perhaps having to maintain a fork. Less work for me!

And yes you can push to my branches I don't care this PR allows maintainers to edit.

**Comment by skarpushin** [comment]
@ctrlcctrlv FYI: I'm still hoping to review & test this during this weekend, but not sure anymore - family reasons. 

BTW, few more questions:

1. Can you please share reason behind "migrate MultipleFilesWatcher to Apache commons I/O"? Current functionality works and I never saw any issues and nobody reported any issues. Why change this functionality and risk to add regression?
2. I vaguely remember reasons for me to create own "patched" version of launch4j `1.7.25-SK` - there was some defect in `1.7.25` that I wanted to be fixed but maintainer of launch4j was not doing it. I tbh don't particularly remember what the defect was. Did you check git log for that change and if new version of launch4j have it fixed?

File: pgptool-gui/installer/launch4j.xml
Comment (skarpushin) [inline pgptool-gui/installer/launch4j.xml:5]: @ctrlcctrlv can you please share -- why you've added version here? Is it matter of "taste" or is it required for something to work?

File: pgptool-gui/installer/launch4j.xml
Comment (skarpushin) [inline pgptool-gui/installer/launch4j.xml:16]: @ctrlcctrlv why removing icon? Can't we just use same approach as above and specify `<icon>${project.build.directory}/icons/favicon.ico</icon>` ?

File: pgptool-gui/installer/launch4j.xml
Comment (ctrlcctrlv) [reply pgptool-gui/installer/launch4j.xml:5]: Required to work on my system.

File: pgptool-gui/installer/launch4j.xml
Comment (ctrlcctrlv) [reply pgptool-gui/installer/launch4j.xml:16]: I did not understand why favicon.ico (`favicon.ico ICO 64x64 64x64+0+0 8-bit sRGB 16958B 0.000u 0:00.000`) was in use. I couldn't find it at the time but located it now. I thought it was a mistake because usually that filename is used for browser tab icon, and I did not recognize the filetype.

I was not able to determine why including it did not work so did remove it. I'll try a bit harder.

```
[ERROR] 
net.sf.launch4j.BuilderException: Icon doesn't exist.
    at net.sf.launch4j.Builder.build (Builder.java:76)
    at com.akathist.maven.plugins.launch4j.Launch4jMojo.doExecute (Launch4jMojo.java:471)
    at com.akathist.maven.plugins.launch4j.Launch4jMojo.execute (Launch4jMojo.java:339)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:298)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:578)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

**Comment by ctrlcctrlv** [comment]
> Can you please share reason behind "migrate MultipleFilesWatcher to Apache commons I/O"? Current functionality works and I never saw any issues and nobody reported any issues. Why change this functionality and risk to add regression?

Because your class is not even uploaded to Maven and I couldn't figure out where it's supposed to come from. Please be aware I don't do much Java development and don't use Windows at all so I came at this from totally different angle to try to get it to build to verify it does build and because I originally wanted to patch #17 and #18.

**Comment by ctrlcctrlv** [comment]
> I vaguely remember reasons for me to create own "patched" version of launch4j 1.7.25-SK - there was some defect in 1.7.25 that I wanted to be fixed but maintainer of launch4j was not doing it. I tbh don't particularly remember what the defect was. Did you check git log for that change and if new version of launch4j have it fixed?

No I did not. Private branches which Maven cannot build lead to a huge mess in general so all private dependencies should be included as submodules or pushed as Maven packages themselves, in my understanding. It's possible the same icon issue I just patched is the one you patched in -SK, but this is conjecture.

**Comment by skarpushin** [comment]
Okay, I thought I've added comment here yesterday, but I don't see it now. 

Rectification of build is blocked by several things -- I'm fixing swingpm and summerb deps. Also there is an issue with launch4j-maven-plugin, I've created an issue https://github.com/orphan-oss/launch4j-maven-plugin/issues/266
I think I'll also submit PR for that fix. "Funny" enough build of launch4j-maven-plugin is also broken (added issue https://github.com/orphan-oss/launch4j-maven-plugin/issues/267)

**Comment by skarpushin** [comment]
Fyi: I've submitted PR for launch4j issue

**Comment by skarpushin** [comment]
@ctrlcctrlv I've just tested "files watching" feature (rework of MultipleFilesWatcher) and it doesn't work. it is not obvious for me from the code perspective what is wrong with it.

File: pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java
Comment (skarpushin) [inline pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:58]: Please define constant for 1000 with self-descriptive name (including units of measurement)

File: pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java
Comment (skarpushin) [inline pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java:101]: Impl of this class and especially of this method doesn't match expectations.

Steps To Reproduce

1. Decrypt fileA.pgp to folderA/fileA
2. Decrypt fileB.pgp to folderA/fileB
3. Use any other application to delete fileA
4. fileA disappears from the "Previously decrypted" files list in PGP Tool
5. Now use any other application to delete fileB

**Actual behavior**: fileB remains in the "Previously decrypted" files list in PGP Tool

**Expected behavior**: fileB must also disappear from the "Previously decrypted" files list in PGP Tool

**How to fix**: you need to track files per folder and stop per-folder-monitors only when no files left to be monitored in this folder.

**Review by skarpushin — CHANGES_REQUESTED** [review]

**Comment by skarpushin** [comment]
@ctrlcctrlv after the fixes above (see comments) I I'm ready _(they accepted my PR for https://github.com/orphan-oss/launch4j-maven-plugin/issues/266 and also I've rectified swingpm dependency)_ to merge these changes into develop, but I want them to be one commit without merge commits. 

**If** you want to this to be done as Merge Request under your name, then can you please create (or update this) PR so that it is based on the tip of `develop` branch and contains only 1 commit (including my edits)?

Otherwise (if you don't care) I can easily incorporate your changes into develop. 

In both cases your name will be in contributors.

Please let me know.

**Comment by skarpushin** [comment]
@ctrlcctrlv that is expected -- they have not yet released fixed version. For now we can live with local version. 
I didn't check exact version, but I think you can do following things

for **launch4j-maven-plugin**
1. clone launch4j-maven-plugin repo locally
2. check version in pom (I think latest one was 2.4.1-SNAPSHOT)
3. run `mvn install` - this will install

then for "PGP Tool"
1. Update pom.xml and set same version for launch4j-maven-plugin that you've installed in local maven repo
2. You're good to go

At least this will be sufficient to run application locally and fix regression

**Comment by ctrlcctrlv** [comment]
Incredible Maven needs hacks like this lol, it tries so hard to pretend to be such an advanced build system but can't even download from Git repo :laughing: 

I added script containing:

```bash
#!/bin/bash
if [[ ! -d "$HOME/.m2/repository/com/akathist/maven/plugins/launch4j" ]]; then
	git clone https://github.com/orphan-oss/launch4j-maven-plugin TEMP-launch4j-mp
	cd TEMP-launch4j-mp
	#VERSION=$(grep '<version>' pom.xml | grep SNAPSHOT | grep -Po '[0-9\.]+-SNAPSHOT')
	mvn install
	cd ..
	rm -rf TEMP-launch4j-mp
fi

VERSION=`ls $HOME/.m2/repository/com/akathist/maven/plugins/launch4j/launch4j-maven-plugin/|grep SNAPSHOT|tee|tail -n 1`
cd pgptool-gui
sed -E ':a;N;$!ba;s@(<artifactId>launch4j-maven-plugin<\/artifactId>[[:space:]]*\n[[:space:]]*<version>)([0-9.]+(-SNAPSHOT)?)</version>@\1'"$VERSION"'<\/version>@m' -i pom.xml
mvn package
```

And a Makefile to call it.

**Comment by skarpushin** [comment]
Well, that's kinda very rare case -- not happening all the time. Since inception of this project -- this is pretty much first occurrence.  Script looks nice but let's not commit such things to repo. I expect fixed version to appear in maven central soon enough.

Are you planning to fix regression with files watcher or you decided to discard it?

**Comment by ctrlcctrlv** [comment]
sorry, there's  a regression in it? i didn't know why you kept discarding it so gave up

**Pull request merged**
